### PR TITLE
RATIS-1479.Make the min gap to take snapshot configurable

### PR DESCRIPTION
**Pull request by codings-dan**
## What changes were proposed in this pull request?

subtask of Support snapshot command: Make the min gap to take snapshot configurable

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/RATIS-1479

## How was this patch tested?
UT

### CODE REVIEW

**Comment by codings-dan** [comment]
The CI test error seems to have nothing to do with this code change.
```
Error:  testMultipleStreamsMultipleServersStepDownLeader(org.apache.ratis.datastream.TestNettyDataStreamChainTopologyWithGrpcCluster)  Time elapsed: 100.012 s  <<< ERROR!
org.junit.runners.model.TestTimedOutException: test timed out after 100 seconds
	at sun.misc.Unsafe.park(Native Method)
	at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
	at java.util.concurrent.FutureTask.awaitDone(FutureTask.java:426)
	at java.util.concurrent.FutureTask.get(FutureTask.java:204)
	at org.junit.internal.runners.statements.FailOnTimeout.getResult(FailOnTimeout.java:141)
	at org.junit.internal.runners.statements.FailOnTimeout.evaluate(FailOnTimeout.java:127)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:298)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:292)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.lang.Thread.run(Thread.java:748)
```

File: ratis-server-api/src/main/java/org/apache/ratis/server/RaftServerConfigKeys.java [hunk 1]
```diff
@@ -433,6 +433,18 @@ static void setAutoTriggerEnabled(RaftProperties properties, boolean autoTrigger
       setBoolean(properties::setBoolean, AUTO_TRIGGER_ENABLED_KEY, autoTriggerThreshold);
     }
 
+    /** min gap to take snapshot async */
+    String MIN_GAP_TAKE_SNAPSHOT_KEY = PREFIX + ".min.gap.takeSnapshot";
```
Comment (szetszwo) [inline ratis-server-api/src/main/java/org/apache/ratis/server/RaftServerConfigKeys.java:437]: Let's call it "creation.gap".

File: ratis-server-api/src/main/java/org/apache/ratis/server/RaftServerConfigKeys.java [hunk 2]
```diff
@@ -433,6 +433,18 @@ static void setAutoTriggerEnabled(RaftProperties properties, boolean autoTrigger
...
+    long MIN_GAP_TAKE_SNAPSHOT_DEFAULT = 5;
```
Comment (szetszwo) [inline ratis-server-api/src/main/java/org/apache/ratis/server/RaftServerConfigKeys.java:438]: The default should be larger.  How about 1024?

File: ratis-server-api/src/main/java/org/apache/ratis/server/RaftServerConfigKeys.java [hunk 3]
```diff
@@ -433,6 +433,18 @@ static void setAutoTriggerEnabled(RaftProperties properties, boolean autoTrigger
...
+    static long minGapTakeSnapshot(RaftProperties properties) {
+      return getLong(
+          properties::getLong, MIN_GAP_TAKE_SNAPSHOT_KEY, MIN_GAP_TAKE_SNAPSHOT_DEFAULT,
+          getDefaultLog(), requireMin(5L));
```
Comment (szetszwo) [inline ratis-server-api/src/main/java/org/apache/ratis/server/RaftServerConfigKeys.java:442]: The min should be 1L.

**Review by szetszwo — COMMENTED** [review]
@codings-dan , thanks for working on this.  Just have some comment on the conf.

```
    /** The log index gap between to two snapshot creations. */
    String CREATION_GAP_KEY = PREFIX + ".creation.gap";
    long CREATION_GAP_DEFAULT = 1024;
    static long creationGap(RaftProperties properties) {
      return getLong(properties::getLong,
          CREATION_GAP_KEY, CREATION_GAP_DEFAULT, getDefaultLog(), requireMin(1L));
    }
    static void setCreationGap(RaftProperties properties, long minGapTakeSnapshot) {
      setLong(properties::setLong, CREATION_GAP_KEY, minGapTakeSnapshot);
    }
```

BTW, we should replace the term "takeSanpshot" with "createSnapshot" since file systems (e.g. HDFS) use the term "create". Also, we may have "creation gap", "creation time", etc.

See if you want to do the rename in this pull request.

**Comment by codings-dan** [comment]
@szetszwo Thanks for reviewing the code, I don’t know much about the term on the file system. In Alluxio, it’s called takesnapshot/checkpoint. I have changed the code according to your comment. In addition, 1024 is really a good number, haha

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Comment by szetszwo** [comment]
> ... In Alluxio, it’s called takesnapshot/checkpoint.  ...

Good to know.  Thanks.

**Pull request merged**